In [95]:
import pandas as pd
import numpy as np
import scipy.stats as stats

In [96]:
df = pd.read_csv("labelled_data/part_D/partd_data.csv")

In [97]:
df

,Unnamed: 0,PRSCRBR_NPI,Prscrbr_Last_Org_Name,Prscrbr_First_Name,Prscrbr_MI,Prscrbr_Crdntls,Prscrbr_Gndr,Prscrbr_Ent_Cd,Prscrbr_St1,Prscrbr_St2,...,Bene_Race_Wht_Cnt,Bene_Race_Black_Cnt,Bene_Race_Api_Cnt,Bene_Race_Hspnc_Cnt,Bene_Race_Natind_Cnt,Bene_Race_Othr_Cnt,Bene_Dual_Cnt,Bene_Ndual_Cnt,Bene_Avg_Risk_Scre,Fraud
0,0,1003000126,Enkeshafi,Ardalan,NaN,M.D.,M,I,6410 Rockledge Dr Ste 304,NaN,...,67.0,27.0,NaN,NaN,0.0,NaN,28.0,78.0,2.245800,0.0
1,1,1003000142,Khalil,Rashid,NaN,M.D.,M,I,4126 N Holland Sylvania Rd,Suite 220,...,130.0,81.0,NaN,12.0,0.0,NaN,125.0,103.0,1.695165,0.0
2,2,1003000167,Escobar,Julio,E,DDS,M,I,5 Pine Cone Rd,NaN,...,41.0,0.0,0.0,NaN,0.0,NaN,0.0,43.0,1.006070,0.0
3,3,1003000175,Reyes-Vasquez,Belinda,NaN,D.D.S.,F,I,322 N Azusa Ave Ste 202,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,1.251869,0.0
4,4,1003000282,Blakemore,Rosie,K,FNP,F,I,Tennessee Prison For Women,3881 Stewarts Lane,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,4.919074,0.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1255119,1255170,1992999569,Yuen,Jenny,F,"O.D., F.A.A.O",F,I,515 Gough St,NaN,...,NaN,0.0,16.0,0.0,0.0,NaN,NaN,NaN,1.451353,0.0
1255120,1255171,1992999650,Yong,Wayne,B,DMD,M,I,28 Atlantic Avenue,121,...,NaN,NaN,18.0,0.0,0.0,NaN,NaN,NaN,1.344691,0.0
1255121,1255172,1992999759,Soriano,Dino,V,RN MSN HSA FNP-C,M,I,171 Clover Point Cir,NaN,...,16.0,NaN,0.0,0.0,0.0,NaN,NaN,NaN,1.545349,1.0
1255122,1255173,1992999825,Deschenes,Geoffrey,R,M.D.,M,I,1100 9th Ave,Ms:m4-Pfs,...,100.0,NaN,16.0,NaN,0.0,NaN,18.0,108.0,1.109445,0.0


In [98]:
df.columns

Index(['Unnamed: 0', 'PRSCRBR_NPI', 'Prscrbr_Last_Org_Name',
       'Prscrbr_First_Name', 'Prscrbr_MI', 'Prscrbr_Crdntls', 'Prscrbr_Gndr',
       'Prscrbr_Ent_Cd', 'Prscrbr_St1', 'Prscrbr_St2', 'Prscrbr_City',
       'Prscrbr_State_Abrvtn', 'Prscrbr_State_FIPS', 'Prscrbr_zip5',
       'Prscrbr_RUCA', 'Prscrbr_RUCA_Desc', 'Prscrbr_Cntry', 'Prscrbr_Type',
       'Prscrbr_Type_src', 'Tot_Clms', 'Tot_30day_Fills', 'Tot_Drug_Cst',
       'Tot_Day_Suply', 'Tot_Benes', 'GE65_Sprsn_Flag', 'GE65_Tot_Clms',
       'GE65_Tot_30day_Fills', 'GE65_Tot_Drug_Cst', 'GE65_Tot_Day_Suply',
       'GE65_Bene_Sprsn_Flag', 'GE65_Tot_Benes', 'Brnd_Sprsn_Flag',
       'Brnd_Tot_Clms', 'Brnd_Tot_Drug_Cst', 'Gnrc_Sprsn_Flag',
       'Gnrc_Tot_Clms', 'Gnrc_Tot_Drug_Cst', 'Othr_Sprsn_Flag',
       'Othr_Tot_Clms', 'Othr_Tot_Drug_Cst', 'MAPD_Sprsn_Flag',
       'MAPD_Tot_Clms', 'MAPD_Tot_Drug_Cst', 'PDP_Sprsn_Flag', 'PDP_Tot_Clms',
       'PDP_Tot_Drug_Cst', 'LIS_Sprsn_Flag', 'LIS_Tot_Clms', 'LIS_Drug_Cst',
       

In [99]:
df["GE65_Tot_Drug_Cst"]

0          17117.40
1          18262.43
2               NaN
3               NaN
4               NaN
             ...   
1255119         NaN
1255120         NaN
1255121     4832.38
1255122    10090.12
1255123    11725.70
Name: GE65_Tot_Drug_Cst, Length: 1255124, dtype: float64

In [100]:
df = df[['PRSCRBR_NPI','Prscrbr_Type','Tot_Clms', 'Tot_30day_Fills', 'Tot_Drug_Cst','Tot_Day_Suply', 'Tot_Benes', 'GE65_Tot_Clms','GE65_Tot_30day_Fills', 
    'GE65_Tot_Drug_Cst', 'GE65_Tot_Day_Suply', 'GE65_Tot_Benes', 'Brnd_Tot_Clms', 'Brnd_Tot_Drug_Cst', 'Gnrc_Tot_Clms', 
    'Gnrc_Tot_Drug_Cst','Othr_Tot_Clms','Othr_Tot_Drug_Cst', 'MAPD_Tot_Clms', 'MAPD_Tot_Drug_Cst', 'PDP_Tot_Clms','PDP_Tot_Drug_Cst', 
    'LIS_Tot_Clms', 'LIS_Drug_Cst','NonLIS_Tot_Clms','NonLIS_Drug_Cst','Opioid_Tot_Clms', 'Opioid_Tot_Drug_Cst', 'Opioid_Tot_Suply',
    'Opioid_Tot_Benes', 'Opioid_LA_Tot_Clms','Opioid_LA_Tot_Drug_Cst', 'Opioid_LA_Tot_Suply', 'Opioid_LA_Tot_Benes','Antbtc_Tot_Clms', 
    'Antbtc_Tot_Drug_Cst','Antbtc_Tot_Benes', 'Antpsyct_GE65_Tot_Clms', 'Antpsyct_GE65_Tot_Drug_Cst','Antpsyct_GE65_Tot_Benes','Fraud']]

In [101]:
df

,PRSCRBR_NPI,Prscrbr_Type,Tot_Clms,Tot_30day_Fills,Tot_Drug_Cst,Tot_Day_Suply,Tot_Benes,GE65_Tot_Clms,GE65_Tot_30day_Fills,GE65_Tot_Drug_Cst,...,Opioid_LA_Tot_Drug_Cst,Opioid_LA_Tot_Suply,Opioid_LA_Tot_Benes,Antbtc_Tot_Clms,Antbtc_Tot_Drug_Cst,Antbtc_Tot_Benes,Antpsyct_GE65_Tot_Clms,Antpsyct_GE65_Tot_Drug_Cst,Antpsyct_GE65_Tot_Benes,Fraud
0,1003000126,Internal Medicine,324,369.800000,20606.08,8621,106.0,277.0,308.233333,17117.40,...,0.00,0.0,NaN,42.0,4868.85,38.0,0.0,0.0,0.0,0.0
1,1003000142,Anesthesiology,1992,2145.666667,79803.65,60953,228.0,727.0,805.066667,18262.43,...,23115.25,4464.0,22.0,NaN,NaN,NaN,0.0,0.0,0.0,0.0
2,1003000167,Dentist,57,57.700000,327.34,554,43.0,NaN,NaN,NaN,...,NaN,NaN,NaN,46.0,262.79,40.0,NaN,NaN,NaN,0.0
3,1003000175,Dentist,18,18.000000,128.52,181,NaN,NaN,NaN,NaN,...,0.00,0.0,NaN,16.0,116.30,NaN,NaN,NaN,NaN,0.0
4,1003000282,Nurse Practitioner,37,47.000000,3834.98,1366,NaN,NaN,NaN,NaN,...,0.00,0.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,0.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1255119,1992999569,Optometry,40,80.666667,9870.01,2357,17.0,NaN,NaN,NaN,...,0.00,0.0,0.0,0.0,0.00,0.0,NaN,NaN,NaN,0.0
1255120,1992999650,Dentist,72,73.200000,397.84,1092,27.0,NaN,NaN,NaN,...,0.00,0.0,0.0,27.0,113.92,18.0,NaN,NaN,NaN,0.0
1255121,1992999759,Nurse Practitioner,47,90.466667,13618.96,2704,21.0,17.0,34.000000,4832.38,...,0.00,0.0,0.0,0.0,0.00,0.0,0.0,0.0,NaN,1.0
1255122,1992999825,Otolaryngology,249,291.066667,10899.52,5801,126.0,217.0,255.066667,10090.12,...,0.00,0.0,0.0,64.0,1233.49,53.0,0.0,0.0,0.0,0.0


In [102]:
df.dtypes

PRSCRBR_NPI                     int64
Prscrbr_Type                   object
Tot_Clms                        int64
Tot_30day_Fills               float64
Tot_Drug_Cst                  float64
Tot_Day_Suply                   int64
Tot_Benes                     float64
GE65_Tot_Clms                 float64
GE65_Tot_30day_Fills          float64
GE65_Tot_Drug_Cst             float64
GE65_Tot_Day_Suply            float64
GE65_Tot_Benes                float64
Brnd_Tot_Clms                 float64
Brnd_Tot_Drug_Cst             float64
Gnrc_Tot_Clms                 float64
Gnrc_Tot_Drug_Cst             float64
Othr_Tot_Clms                 float64
Othr_Tot_Drug_Cst             float64
MAPD_Tot_Clms                 float64
MAPD_Tot_Drug_Cst             float64
PDP_Tot_Clms                  float64
PDP_Tot_Drug_Cst              float64
LIS_Tot_Clms                  float64
LIS_Drug_Cst                  float64
NonLIS_Tot_Clms               float64
NonLIS_Drug_Cst               float64
Opioid_Tot_C

In [103]:
print(df["Tot_Clms"].isnull().sum())
print(df["Tot_30day_Fills"].isnull().sum())
print(df["Tot_Drug_Cst"].isnull().sum())
print(df["Tot_Day_Suply"].isnull().sum())
print("-------------------------------------------------")
print(df["Tot_Benes"].isnull().sum())
print("-------------------------------------------------")
print(df["GE65_Tot_Clms"].isnull().sum())
print(df["GE65_Tot_30day_Fills"].isnull().sum())
print(df["GE65_Tot_Drug_Cst"].isnull().sum())
print(df["GE65_Tot_Day_Suply"].isnull().sum())
print(df["GE65_Tot_Benes"].isnull().sum())
print("-------------------------------------------------")
print(df["Brnd_Tot_Clms"].isnull().sum())
print(df["Brnd_Tot_Drug_Cst"].isnull().sum())
print("-------------------------------------------------")
print(df["Gnrc_Tot_Clms"].isnull().sum())
print(df["Gnrc_Tot_Drug_Cst"].isnull().sum())
print("-------------------------------------------------")
print(df["Othr_Tot_Clms"].isnull().sum())
print(df["Othr_Tot_Drug_Cst"].isnull().sum())
print("-------------------------------------------------")
print(df["MAPD_Tot_Clms"].isnull().sum())
print(df["MAPD_Tot_Drug_Cst"].isnull().sum())
print("-------------------------------------------------")
print(df["PDP_Tot_Clms"].isnull().sum())
print(df["PDP_Tot_Drug_Cst"].isnull().sum())
print(df["LIS_Tot_Clms"].isnull().sum())
print(df["LIS_Drug_Cst"].isnull().sum())
print("-------------------------------------------------")
print(df["Opioid_Tot_Clms"].isnull().sum())
print(df["Opioid_Tot_Drug_Cst"].isnull().sum())
print(df["Opioid_Tot_Suply"].isnull().sum())
print(df["Opioid_Tot_Benes"].isnull().sum())
print("-------------------------------------------------")
print(df["Opioid_LA_Tot_Clms"].isnull().sum())
print(df["Opioid_LA_Tot_Drug_Cst"].isnull().sum())
print(df["Opioid_LA_Tot_Suply"].isnull().sum())
print(df["Opioid_LA_Tot_Benes"].isnull().sum())
print("-------------------------------------------------")
print(df["Antbtc_Tot_Clms"].isnull().sum())
print(df["Antbtc_Tot_Drug_Cst"].isnull().sum())
print(df["Antbtc_Tot_Benes"].isnull().sum())
print("-------------------------------------------------")
print(df["Antpsyct_GE65_Tot_Clms"].isnull().sum())
print(df["Antpsyct_GE65_Tot_Drug_Cst"].isnull().sum())
print(df["Antpsyct_GE65_Tot_Benes"].isnull().sum())
print("-------------------------------------------------")
print(df["Opioid_LA_Tot_Benes"].isnull().sum())
print(df["Opioid_LA_Tot_Benes"].isnull().sum())
print(df["Opioid_LA_Tot_Benes"].isnull().sum())
print(df["Opioid_LA_Tot_Benes"].isnull().sum())

0
0
0
0
-------------------------------------------------
142956
-------------------------------------------------
294227
294227
294227
294227
592000
-------------------------------------------------
543484
543484
-------------------------------------------------
30172
30172
-------------------------------------------------
538251
538251
-------------------------------------------------
231297
231297
-------------------------------------------------
231297
231297
266730
266730
-------------------------------------------------
347133
347133
347133
523631
-------------------------------------------------
432770
432770
432770
651650
-------------------------------------------------
373592
373592
501293
-------------------------------------------------
459981
459981
797381
-------------------------------------------------
651650
651650
651650
651650


## Finding where all the major columns are NULL

In [104]:
x = len(df)
y = len(df.columns)
all_nan1 = []
for i in range(x):
    if df.iloc[i].isnull().sum()>=34:
        all_nan1.append(i)

In [105]:
len(all_nan1)

121

## Dropping all 121 records from the dataframe and all those records are non-fraud

In [106]:
all_nan_df = []
for i in all_nan1:
    df = df.drop(i,axis=0)
    all_nan_df.append(df.iloc[i])

In [107]:
 pd.DataFrame(all_nan_df)

,PRSCRBR_NPI,Prscrbr_Type,Tot_Clms,Tot_30day_Fills,Tot_Drug_Cst,Tot_Day_Suply,Tot_Benes,GE65_Tot_Clms,GE65_Tot_30day_Fills,GE65_Tot_Drug_Cst,...,Opioid_LA_Tot_Drug_Cst,Opioid_LA_Tot_Suply,Opioid_LA_Tot_Benes,Antbtc_Tot_Clms,Antbtc_Tot_Drug_Cst,Antbtc_Tot_Benes,Antpsyct_GE65_Tot_Clms,Antpsyct_GE65_Tot_Drug_Cst,Antpsyct_GE65_Tot_Benes,Fraud
8473,1003840505,Obstetrics & Gynecology,107,158.766667,1536.94,4722,15.0,94.0,145.766667,1498.90,...,0.0,0.0,0.0,0.0,0.00,0.0,0.0,0.00,NaN,0.0
33755,1023397544,Dentist,75,92.000000,412.53,1102,42.0,NaN,NaN,NaN,...,0.0,0.0,NaN,35.0,148.43,30.0,NaN,NaN,NaN,0.0
40698,1033188453,Dentist,65,65.000000,294.42,540,46.0,NaN,NaN,NaN,...,0.0,0.0,0.0,65.0,294.42,46.0,NaN,NaN,NaN,0.0
54403,1043326911,Psychiatry,1488,1603.800000,222316.39,47880,48.0,624.0,726.400000,51801.06,...,0.0,0.0,0.0,NaN,NaN,NaN,146.0,42219.57,14.0,0.0
73843,1053852566,Student in an Organized Health Care Education/...,109,109.000000,793.29,668,51.0,91.0,91.000000,668.74,...,0.0,0.0,0.0,46.0,485.97,44.0,0.0,0.00,NaN,0.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1199409,1952601312,Podiatry,239,276.333333,3611.69,6320,92.0,174.0,190.666667,2465.37,...,0.0,0.0,0.0,23.0,400.35,13.0,0.0,0.00,0.0,0.0
1212370,1962672584,Family Practice,3281,7832.866667,220933.79,230707,348.0,2944.0,7143.966667,192436.44,...,0.0,0.0,0.0,56.0,1392.91,45.0,11.0,386.12,NaN,0.0
1232399,1982149878,Nurse Practitioner,127,129.000000,4264.12,1633,91.0,98.0,98.000000,3645.38,...,NaN,NaN,NaN,56.0,502.84,53.0,0.0,0.00,0.0,0.0
1245075,1992157606,Nurse Practitioner,3645,8509.766667,286874.34,248686,366.0,3276.0,7735.566667,276068.49,...,0.0,0.0,0.0,90.0,968.41,56.0,NaN,NaN,NaN,0.0


In [108]:
all_nan_df = pd.DataFrame(all_nan_df)

In [109]:
all_nan_df.iloc[2].isnull().sum()

12

In [110]:
all_nan_df["Fraud"].value_counts()

0.0    121
Name: Fraud, dtype: int64

In [111]:
df

,PRSCRBR_NPI,Prscrbr_Type,Tot_Clms,Tot_30day_Fills,Tot_Drug_Cst,Tot_Day_Suply,Tot_Benes,GE65_Tot_Clms,GE65_Tot_30day_Fills,GE65_Tot_Drug_Cst,...,Opioid_LA_Tot_Drug_Cst,Opioid_LA_Tot_Suply,Opioid_LA_Tot_Benes,Antbtc_Tot_Clms,Antbtc_Tot_Drug_Cst,Antbtc_Tot_Benes,Antpsyct_GE65_Tot_Clms,Antpsyct_GE65_Tot_Drug_Cst,Antpsyct_GE65_Tot_Benes,Fraud
0,1003000126,Internal Medicine,324,369.800000,20606.08,8621,106.0,277.0,308.233333,17117.40,...,0.00,0.0,NaN,42.0,4868.85,38.0,0.0,0.0,0.0,0.0
1,1003000142,Anesthesiology,1992,2145.666667,79803.65,60953,228.0,727.0,805.066667,18262.43,...,23115.25,4464.0,22.0,NaN,NaN,NaN,0.0,0.0,0.0,0.0
2,1003000167,Dentist,57,57.700000,327.34,554,43.0,NaN,NaN,NaN,...,NaN,NaN,NaN,46.0,262.79,40.0,NaN,NaN,NaN,0.0
3,1003000175,Dentist,18,18.000000,128.52,181,NaN,NaN,NaN,NaN,...,0.00,0.0,NaN,16.0,116.30,NaN,NaN,NaN,NaN,0.0
4,1003000282,Nurse Practitioner,37,47.000000,3834.98,1366,NaN,NaN,NaN,NaN,...,0.00,0.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,0.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1255119,1992999569,Optometry,40,80.666667,9870.01,2357,17.0,NaN,NaN,NaN,...,0.00,0.0,0.0,0.0,0.00,0.0,NaN,NaN,NaN,0.0
1255120,1992999650,Dentist,72,73.200000,397.84,1092,27.0,NaN,NaN,NaN,...,0.00,0.0,0.0,27.0,113.92,18.0,NaN,NaN,NaN,0.0
1255121,1992999759,Nurse Practitioner,47,90.466667,13618.96,2704,21.0,17.0,34.000000,4832.38,...,0.00,0.0,0.0,0.0,0.00,0.0,0.0,0.0,NaN,1.0
1255122,1992999825,Otolaryngology,249,291.066667,10899.52,5801,126.0,217.0,255.066667,10090.12,...,0.00,0.0,0.0,64.0,1233.49,53.0,0.0,0.0,0.0,0.0


In [112]:
print(df["Tot_Clms"].isnull().sum())
print(df["Tot_30day_Fills"].isnull().sum())
print(df["Tot_Drug_Cst"].isnull().sum())
print(df["Tot_Day_Suply"].isnull().sum())
print("-------------------------------------------------")
print(df["Tot_Benes"].isnull().sum())
print("-------------------------------------------------")
print(df["GE65_Tot_Clms"].isnull().sum())
print(df["GE65_Tot_30day_Fills"].isnull().sum())
print(df["GE65_Tot_Drug_Cst"].isnull().sum())
print(df["GE65_Tot_Day_Suply"].isnull().sum())
print(df["GE65_Tot_Benes"].isnull().sum())
print("-------------------------------------------------")
print(df["Brnd_Tot_Clms"].isnull().sum())
print(df["Brnd_Tot_Drug_Cst"].isnull().sum())
print("-------------------------------------------------")
print(df["Gnrc_Tot_Clms"].isnull().sum())
print(df["Gnrc_Tot_Drug_Cst"].isnull().sum())
print("-------------------------------------------------")
print(df["Othr_Tot_Clms"].isnull().sum())
print(df["Othr_Tot_Drug_Cst"].isnull().sum())
print("-------------------------------------------------")
print(df["MAPD_Tot_Clms"].isnull().sum())
print(df["MAPD_Tot_Drug_Cst"].isnull().sum())
print("-------------------------------------------------")
print(df["PDP_Tot_Clms"].isnull().sum())
print(df["PDP_Tot_Drug_Cst"].isnull().sum())
print(df["LIS_Tot_Clms"].isnull().sum())
print(df["LIS_Drug_Cst"].isnull().sum())
print("-------------------------------------------------")
print(df["Opioid_Tot_Clms"].isnull().sum())
print(df["Opioid_Tot_Drug_Cst"].isnull().sum())
print(df["Opioid_Tot_Suply"].isnull().sum())
print(df["Opioid_Tot_Benes"].isnull().sum())
print("-------------------------------------------------")
print(df["Opioid_LA_Tot_Clms"].isnull().sum())
print(df["Opioid_LA_Tot_Drug_Cst"].isnull().sum())
print(df["Opioid_LA_Tot_Suply"].isnull().sum())
print(df["Opioid_LA_Tot_Benes"].isnull().sum())
print("-------------------------------------------------")
print(df["Antbtc_Tot_Clms"].isnull().sum())
print(df["Antbtc_Tot_Drug_Cst"].isnull().sum())
print(df["Antbtc_Tot_Benes"].isnull().sum())
print("-------------------------------------------------")
print(df["Antpsyct_GE65_Tot_Clms"].isnull().sum())
print(df["Antpsyct_GE65_Tot_Drug_Cst"].isnull().sum())
print(df["Antpsyct_GE65_Tot_Benes"].isnull().sum())
print("-------------------------------------------------")
print(df["Opioid_LA_Tot_Benes"].isnull().sum())
print(df["Opioid_LA_Tot_Benes"].isnull().sum())
print(df["Opioid_LA_Tot_Benes"].isnull().sum())
print(df["Opioid_LA_Tot_Benes"].isnull().sum())

0
0
0
0
-------------------------------------------------
142835
-------------------------------------------------
294106
294106
294106
294106
591879
-------------------------------------------------
543363
543363
-------------------------------------------------
30051
30051
-------------------------------------------------
538130
538130
-------------------------------------------------
231176
231176
-------------------------------------------------
231176
231176
266609
266609
-------------------------------------------------
347012
347012
347012
523510
-------------------------------------------------
432649
432649
432649
651529
-------------------------------------------------
373471
373471
501172
-------------------------------------------------
459860
459860
797260
-------------------------------------------------
651529
651529
651529
651529


## Now, we divide accordingly

## First GDRC coz we have less

In [113]:
print(df["Gnrc_Tot_Clms"].isnull().sum())
print(df["Gnrc_Tot_Drug_Cst"].isnull().sum())

30051
30051


In [118]:
gnrc_nan = []
no_gnrc = []
for i in range(len(df)):
    if(pd.isnull(df.iloc[i].Gnrc_Tot_Clms) == True and pd.isnull(df.iloc[i].Gnrc_Tot_Drug_Cst)):
        gnrc_nan.append(df.iloc[i])
    else:
        no_gnrc.append(df.iloc[i])

In [119]:
gnrc_nan = pd.DataFrame(gnrc_nan)
no_gnrc = pd.DataFrame(no_gnrc)

In [120]:
print(len(gnrc_nan))
gnrc_nan["Fraud"].value_counts()

30051


0.0    30001
1.0       50
Name: Fraud, dtype: int64

In [62]:
df.drop("Prscrbr_Type",axis=1).corr()

,PRSCRBR_NPI,Tot_Clms,Tot_30day_Fills,Tot_Drug_Cst,Tot_Day_Suply,Tot_Benes,GE65_Tot_Clms,GE65_Tot_30day_Fills,GE65_Tot_Drug_Cst,GE65_Tot_Day_Suply,...,Opioid_LA_Tot_Drug_Cst,Opioid_LA_Tot_Suply,Opioid_LA_Tot_Benes,Antbtc_Tot_Clms,Antbtc_Tot_Drug_Cst,Antbtc_Tot_Benes,Antpsyct_GE65_Tot_Clms,Antpsyct_GE65_Tot_Drug_Cst,Antpsyct_GE65_Tot_Benes,Fraud
PRSCRBR_NPI,1.000000,-0.000505,-0.000354,0.000972,-0.000343,0.001489,-0.000481,-0.000356,0.001504,-0.000353,...,-0.001128,-0.001109,-0.001050,-0.001270,0.000135,-0.001314,-0.000835,-0.000176,0.001056,-0.000349
Tot_Clms,-0.000505,1.000000,0.953939,0.543884,0.919410,0.501186,0.980677,0.923330,0.537388,0.886105,...,0.160435,0.208818,0.193927,0.645953,0.154193,0.550095,0.546575,0.243171,0.594437,0.005485
Tot_30day_Fills,-0.000354,0.953939,1.000000,0.521138,0.991127,0.476152,0.951925,0.989000,0.528775,0.978563,...,0.117490,0.156394,0.126815,0.619078,0.126023,0.555469,0.398686,0.169774,0.486101,0.005303
Tot_Drug_Cst,0.000972,0.543884,0.521138,1.000000,0.508296,0.373992,0.502474,0.476178,0.943861,0.461423,...,0.105978,0.121170,0.114902,0.335499,0.192723,0.258902,0.250240,0.215825,0.268330,0.002814
Tot_Day_Suply,-0.000343,0.919410,0.991127,0.508296,1.000000,0.407318,0.914968,0.980719,0.513159,0.989076,...,0.116134,0.153680,0.122048,0.596029,0.114624,0.541616,0.344687,0.154459,0.451857,0.005189
Tot_Benes,0.001489,0.501186,0.476152,0.373992,0.407318,1.000000,0.499381,0.460563,0.397717,0.384599,...,0.103209,0.136292,0.161214,0.284966,0.062034,0.296099,0.167091,0.079316,0.227059,0.003204
GE65_Tot_Clms,-0.000481,0.980677,0.951925,0.502474,0.914968,0.499381,1.000000,0.951742,0.548354,0.913383,...,0.105431,0.146279,0.125488,0.636820,0.138592,0.549454,0.489162,0.187980,0.533988,0.007063
GE65_Tot_30day_Fills,-0.000356,0.923330,0.989000,0.476178,0.980719,0.460563,0.951742,1.000000,0.528051,0.990446,...,0.076087,0.108678,0.074641,0.598715,0.108121,0.544935,0.349235,0.128154,0.429935,0.006877
GE65_Tot_Drug_Cst,0.001504,0.537388,0.528775,0.943861,0.513159,0.397717,0.548354,0.528051,1.000000,0.510923,...,0.068530,0.086444,0.080088,0.330083,0.152052,0.265880,0.205519,0.154632,0.225266,0.003844
GE65_Tot_Day_Suply,-0.000353,0.886105,0.978563,0.461423,0.989076,0.384599,0.913383,0.990446,0.510923,1.000000,...,0.074412,0.105461,0.069374,0.575124,0.096837,0.530818,0.295660,0.113168,0.394002,0.006726


In [121]:
gnrc_nan

,PRSCRBR_NPI,Prscrbr_Type,Tot_Clms,Tot_30day_Fills,Tot_Drug_Cst,Tot_Day_Suply,Tot_Benes,GE65_Tot_Clms,GE65_Tot_30day_Fills,GE65_Tot_Drug_Cst,...,Opioid_LA_Tot_Drug_Cst,Opioid_LA_Tot_Suply,Opioid_LA_Tot_Benes,Antbtc_Tot_Clms,Antbtc_Tot_Drug_Cst,Antbtc_Tot_Benes,Antpsyct_GE65_Tot_Clms,Antpsyct_GE65_Tot_Drug_Cst,Antpsyct_GE65_Tot_Benes,Fraud
55,1003004128,Internal Medicine,43,59.933333,15649.20,1766,NaN,24.0,26.266667,8897.99,...,0.0,0.0,NaN,0.0,0.00,NaN,0.0,0.0,NaN,0.0
72,1003005745,Neurology,23,25.000000,1038222.89,672,NaN,0.0,0.000000,0.00,...,0.0,0.0,NaN,0.0,0.00,NaN,0.0,0.0,NaN,0.0
181,1003012022,Pulmonary Disease,169,204.766667,47070.16,5232,55.0,101.0,113.000000,28566.33,...,0.0,0.0,0.0,27.0,3176.55,19.0,0.0,0.0,0.0,0.0
217,1003013301,Dentist,11,13.600000,111.85,224,NaN,NaN,NaN,NaN,...,0.0,0.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,0.0
273,1003015553,Diagnostic Radiology,12,12.000000,157.81,343,NaN,12.0,12.000000,157.81,...,0.0,0.0,NaN,0.0,0.00,NaN,0.0,0.0,NaN,0.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1254827,1992979322,Physician Assistant,11,11.000000,2381.37,210,NaN,11.0,11.000000,2381.37,...,0.0,0.0,NaN,NaN,NaN,NaN,0.0,0.0,NaN,0.0
1254956,1992989271,Dentist,15,15.000000,423.66,226,11.0,15.0,15.000000,423.66,...,0.0,0.0,0.0,NaN,NaN,NaN,0.0,0.0,0.0,0.0
1254973,1992990360,Optometry,12,28.666667,318.80,832,NaN,12.0,28.666667,318.80,...,0.0,0.0,NaN,0.0,0.00,NaN,0.0,0.0,NaN,0.0
1254995,1992991335,Specialist,12,24.900000,10649.70,747,NaN,0.0,0.000000,0.00,...,0.0,0.0,NaN,0.0,0.00,NaN,0.0,0.0,NaN,0.0


In [123]:
print(len(no_gnrc))
no_gnrc["Fraud"].value_counts()

1224952


0.0    1223508
1.0       1444
Name: Fraud, dtype: int64

In [126]:
gnrc_nan.drop(["Gnrc_Tot_Clms","Gnrc_Tot_Drug_Cst"],axis=1)

,PRSCRBR_NPI,Prscrbr_Type,Tot_Clms,Tot_30day_Fills,Tot_Drug_Cst,Tot_Day_Suply,Tot_Benes,GE65_Tot_Clms,GE65_Tot_30day_Fills,GE65_Tot_Drug_Cst,...,Opioid_LA_Tot_Drug_Cst,Opioid_LA_Tot_Suply,Opioid_LA_Tot_Benes,Antbtc_Tot_Clms,Antbtc_Tot_Drug_Cst,Antbtc_Tot_Benes,Antpsyct_GE65_Tot_Clms,Antpsyct_GE65_Tot_Drug_Cst,Antpsyct_GE65_Tot_Benes,Fraud
55,1003004128,Internal Medicine,43,59.933333,15649.20,1766,NaN,24.0,26.266667,8897.99,...,0.0,0.0,NaN,0.0,0.00,NaN,0.0,0.0,NaN,0.0
72,1003005745,Neurology,23,25.000000,1038222.89,672,NaN,0.0,0.000000,0.00,...,0.0,0.0,NaN,0.0,0.00,NaN,0.0,0.0,NaN,0.0
181,1003012022,Pulmonary Disease,169,204.766667,47070.16,5232,55.0,101.0,113.000000,28566.33,...,0.0,0.0,0.0,27.0,3176.55,19.0,0.0,0.0,0.0,0.0
217,1003013301,Dentist,11,13.600000,111.85,224,NaN,NaN,NaN,NaN,...,0.0,0.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,0.0
273,1003015553,Diagnostic Radiology,12,12.000000,157.81,343,NaN,12.0,12.000000,157.81,...,0.0,0.0,NaN,0.0,0.00,NaN,0.0,0.0,NaN,0.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1254827,1992979322,Physician Assistant,11,11.000000,2381.37,210,NaN,11.0,11.000000,2381.37,...,0.0,0.0,NaN,NaN,NaN,NaN,0.0,0.0,NaN,0.0
1254956,1992989271,Dentist,15,15.000000,423.66,226,11.0,15.0,15.000000,423.66,...,0.0,0.0,0.0,NaN,NaN,NaN,0.0,0.0,0.0,0.0
1254973,1992990360,Optometry,12,28.666667,318.80,832,NaN,12.0,28.666667,318.80,...,0.0,0.0,NaN,0.0,0.00,NaN,0.0,0.0,NaN,0.0
1254995,1992991335,Specialist,12,24.900000,10649.70,747,NaN,0.0,0.000000,0.00,...,0.0,0.0,NaN,0.0,0.00,NaN,0.0,0.0,NaN,0.0


In [127]:
gnrc_nan.columns

Index(['PRSCRBR_NPI', 'Prscrbr_Type', 'Tot_Clms', 'Tot_30day_Fills',
       'Tot_Drug_Cst', 'Tot_Day_Suply', 'Tot_Benes', 'GE65_Tot_Clms',
       'GE65_Tot_30day_Fills', 'GE65_Tot_Drug_Cst', 'GE65_Tot_Day_Suply',
       'GE65_Tot_Benes', 'Brnd_Tot_Clms', 'Brnd_Tot_Drug_Cst', 'Gnrc_Tot_Clms',
       'Gnrc_Tot_Drug_Cst', 'Othr_Tot_Clms', 'Othr_Tot_Drug_Cst',
       'MAPD_Tot_Clms', 'MAPD_Tot_Drug_Cst', 'PDP_Tot_Clms',
       'PDP_Tot_Drug_Cst', 'LIS_Tot_Clms', 'LIS_Drug_Cst', 'NonLIS_Tot_Clms',
       'NonLIS_Drug_Cst', 'Opioid_Tot_Clms', 'Opioid_Tot_Drug_Cst',
       'Opioid_Tot_Suply', 'Opioid_Tot_Benes', 'Opioid_LA_Tot_Clms',
       'Opioid_LA_Tot_Drug_Cst', 'Opioid_LA_Tot_Suply', 'Opioid_LA_Tot_Benes',
       'Antbtc_Tot_Clms', 'Antbtc_Tot_Drug_Cst', 'Antbtc_Tot_Benes',
       'Antpsyct_GE65_Tot_Clms', 'Antpsyct_GE65_Tot_Drug_Cst',
       'Antpsyct_GE65_Tot_Benes', 'Fraud'],
      dtype='object')

In [149]:
gnrc_nan["Opioid_LA_Tot_Benes"].value_counts()

0.0    6666
Name: Opioid_LA_Tot_Benes, dtype: int64

In [150]:
gnrc_nan = gnrc_nan.drop(["Gnrc_Tot_Clms","Gnrc_Tot_Drug_Cst"],axis=1)

In [152]:
gnrc_nan.to_csv("labelled_data/part_D/no_gnrc.csv")

In [26]:
fraud_tot_ben = []
nf_tot_ben = []
for i in range(len(df)):
    if(pd.isnull(df.iloc[i].Tot_Benes)== True):
        if(df.iloc[i].Fraud == 1):
            fraud_tot_ben.append(i)
        else:
            nf_tot_ben.append(i)

In [28]:
len(fraud_tot_ben)

243

In [29]:
len(nf_tot_ben)

142592

In [31]:
fraud_tot = []
nf_tot = []
for i in range(len(df)):
    if pd.isnull(df.iloc[i].Tot_Clms)== True and pd.isnull(df.iloc[i].Tot_30day_Fills)== True and pd.isnull(df.iloc[i].Tot_Drug_Cst)== True and pd.isnull(df.iloc[i].Tot_Day_Suply)== True and pd.isnull(df.iloc[i].Tot_Benes)== True:
        if(df.iloc[i].Fraud == 1):
            fraud_tot.append(i)
        else:
            nf_tot.append(i)

In [32]:
len(fraud_tot)

0

In [33]:
len(nf_tot)

0

## We see that the Tot_Benes is inconsistent with Tot so, we now check with every other column

### With Tot_Benes

In [36]:
fraud_ge6 = []
nf_ge6 = []
for i in range(len(df)):
    if pd.isnull(df.iloc[i].GE65_Tot_Clms)== True and pd.isnull(df.iloc[i].GE65_Tot_30day_Fills)== True and pd.isnull(df.iloc[i].GE65_Tot_Drug_Cst)== True and pd.isnull(df.iloc[i].GE65_Tot_Day_Suply)== True and pd.isnull(df.iloc[i].GE65_Tot_Benes)== True and pd.isnull(df.iloc[i].Tot_Benes):
        if(df.iloc[i].Fraud == 1):
            fraud_ge6.append(i)
        else:
            nf_ge6.append(i)

In [38]:
print(len(fraud_ge6))
print(len(nf_ge6))

97
66561


### Without Tot_Benes

In [39]:
fraud_ge6_wt = []
nf_ge6_wt = []
for i in range(len(df)):
    if pd.isnull(df.iloc[i].GE65_Tot_Clms)== True and pd.isnull(df.iloc[i].GE65_Tot_30day_Fills)== True and pd.isnull(df.iloc[i].GE65_Tot_Drug_Cst)== True and pd.isnull(df.iloc[i].GE65_Tot_Day_Suply)== True and pd.isnull(df.iloc[i].GE65_Tot_Benes)== True:
        if(df.iloc[i].Fraud == 1):
            fraud_ge6_wt.append(i)
        else:
            nf_ge6_wt.append(i)

In [40]:
print(len(fraud_ge6_wt))
print(len(nf_ge6_wt))

384
293722


In [58]:
ge6 = []
for i in range(len(df)):
    if pd.isnull(df.iloc[i].GE65_Tot_Clms)== True and pd.isnull(df.iloc[i].GE65_Tot_30day_Fills)== True and pd.isnull(df.iloc[i].GE65_Tot_Drug_Cst)== True and pd.isnull(df.iloc[i].GE65_Tot_Day_Suply)== True and pd.isnull(df.iloc[i].GE65_Tot_Benes)and pd.isnull(df.iloc[i].Tot_Benes)== True:
        ge6.append(i)

In [57]:
selected_columns = ["Tot_Benes","GE65_Tot_Clms","GE65_Tot_30day_Fills","GE65_Tot_Drug_Cst"]
df1_ge65 = df.drop(columns = selected_columns).copy().head(0)
df1_ge65 = pd.DataFrame(df1_ge65)
df1_ge65

,PRSCRBR_NPI,Prscrbr_Type,Tot_Clms,Tot_30day_Fills,Tot_Drug_Cst,Tot_Day_Suply,GE65_Tot_Day_Suply,GE65_Tot_Benes,Brnd_Tot_Clms,Brnd_Tot_Drug_Cst,...,Opioid_LA_Tot_Drug_Cst,Opioid_LA_Tot_Suply,Opioid_LA_Tot_Benes,Antbtc_Tot_Clms,Antbtc_Tot_Drug_Cst,Antbtc_Tot_Benes,Antpsyct_GE65_Tot_Clms,Antpsyct_GE65_Tot_Drug_Cst,Antpsyct_GE65_Tot_Benes,Fraud


In [ ]:
for i in ge6:
    